File: src/main/java/xf/xflp/base/container/Container.java
package xf.xflp.base.container;

import xf.xflp.base.item.Item;
import xf.xflp.base.item.Position;

import java.util.List;

/**
 * Copyright (c) 2012-2022 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @author hschneid
 */
public interface Container {

    /**
     *  Creates a new container instance. The master data of the called container are initialized.
     *  The new container is empty.
     */
    Container newInstance();

    /**
     * Adds an item at given position to current container.
     * New positions will be created and covered positions get inactive.
     *
     * Returns the index of the inserted item for faster access
     */
    int add(Item item, Position position, boolean isRotated);

    /**
     * Removes the item from container.
     * Uncovered positions will be freed.
     */
    void remove(Item item);

    /**
     * Returns the already inserted items of this container
     */
    List<Item> getItems();

    /**
     * History when and how a certain item is loaded or unloaded
     * This is relevant for creating the solution report
     **/
    List<Item> getHistory();

    /**
     * Check if the given item is allowed to be placed at this container
     */
    boolean isItemAllowed(Item item);

    /**
     * Returns the list of active positions, which can be used to add an item here.
     */
    List<Position> getActivePositions();

    /**
     * Returns the type of the container, which is necessary for multi-container-type problems
     */
    int getContainerType();

    /**
     * Returns the width of the container
     */
    int getWidth();

    /**
     * Returns the length of the container
     */
    int getLength();

    /**
     * Returns the height of the container
     */
    int getHeight();

    /**
     * Returns the maximal weight, which can be loaded by the container
     */
    float getMaxWeight();

    /**
     * Returns the loaded volume (volume of inserted items)
     */
    long getLoadedVolume();

    /**
     * Returns the loaded weight (weight of inserted items)
     */
    float getLoadedWeight();

    /**
     * Returns the paramter for behaviour of this container
     */
    ContainerParameter getParameter();

    /**
     * Returns internal data structures of container
     */
    ContainerBaseData getBaseData();
}

File: src/xflp/base/container/container.py
"""
Container interface protocol.

Copyright (c) 2012-2025 Holger Schneider
Licensed under the MIT License
"""

from typing import Protocol, List, TYPE_CHECKING

if TYPE_CHECKING:
    from ..item.item import Item
    from ..item.position import Position
    from ..item.space import Space
    from .container_parameter import ContainerParameter
    from .container_base_data import ContainerBaseData


class Container(Protocol):
    """
    Protocol defining the container interface.
    
    A container is a space where items can be loaded and unloaded.
    It manages positions, validates constraints, and tracks loaded items.
    """
    
    def new_instance(self) -> 'Container':
        """
        Create a new container instance with same master data.
        
        The new container is empty.
        
        Returns:
            A new empty container instance
        """
        ...
    
    def add(self, item: 'Item', position: 'Position', is_rotated: bool) -> int:
        """
        Add an item at given position to container.
        
        New positions will be created and covered positions become inactive.
        
        Args:
            item: The item to add
            position: The position to place the item
            is_rotated: Whether the item is rotated
            
        Returns:
            The index of the inserted item for faster access
        """
        ...
    
    def remove(self, item: 'Item') -> None:
        """
        Remove the item from container.
        
        Uncovered positions will be freed.
        
        Args:
            item: The item to remove
        """
        ...
    
    def get_items(self) -> List['Item']:
        """
        Get the already inserted items of this container.
        
        Returns:
            List of items in the container
        """
        ...
    
    def get_history(self) -> List['Item']:
        """
        Get the history of loaded/unloaded items.
        
        This is relevant for creating the solution report.
        
        Returns:
            List of items in order of loading/unloading
        """
        ...
    
    def is_item_allowed(self, item: 'Item') -> bool:
        """
        Check if the given item is allowed to be placed in this container.
        
        Args:
            item: The item to check
            
        Returns:
            True if item can be placed in this container
        """
        ...
    
    def get_active_positions(self) -> List['Position']:
        """
        Get the list of active positions where items can be added.
        
        Returns:
            List of active positions
        """
        ...
    
    def get_container_type(self) -> int:
        """
        Get the type of the container.
        
        Necessary for multi-container-type problems.
        
        Returns:
            The container type ID
        """
        ...
    
    def get_width(self) -> int:
        """Get the width of the container."""
        ...
    
    def get_length(self) -> int:
        """Get the length of the container."""
        ...
    
    def get_height(self) -> int:
        """Get the height of the container."""
        ...
    
    def get_max_weight(self) -> float:
        """Get the maximum weight the container can hold."""
        ...
    
    def get_loaded_volume(self) -> int:
        """Get the loaded volume (volume of inserted items)."""
        ...
    
    def get_loaded_weight(self) -> float:
        """Get the loaded weight (weight of inserted items)."""
        ...
    
    def get_parameter(self) -> 'ContainerParameter':
        """Get the parameters for behavior of this container."""
        ...
    
    def get_base_data(self) -> 'ContainerBaseData':
        """Get internal data structures of container."""
        ...
    
    def get_space(self, pos: 'Position') -> List['Space']:
        """
        Get available spaces at a position.
        
        Args:
            pos: The position to check
            
        Returns:
            List of available spaces
        """
        ...
